flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/sinks/TableSinkUtils.scala

```diff
@@ -189,13 +190,45 @@ object TableSinkUtils {
       queryLogicalType,
       withChangeFlag)
     if (LogicalTypeChecks.isCompositeType(requestedOutputType.getLogicalType)) {
-      DataTypeUtils.expandCompositeTypeToSchema(requestedOutputType)
+      fromDataTypeToTypeInfo(requestedOutputType) match {
+        case pj: PojoTypeInfo[_] => expandPojoTypeToSchema(pj, queryLogicalType)
+        case _ => DataTypeUtils.expandCompositeTypeToSchema(requestedOutputType)
+      }
     } else {
       // atomic type
       TableSchema.builder().field("f0", requestedOutputType).build()
     }
   }
 
+  /**
```

Comment:
Can you extract pojo mapping in `SinkCodeGenerator` to here too? Both are for POJO reorder.

Reply:
There is not too much code can be shared (4 lines) and the logic is different: `SinkCodeGenerator` extracts a fields mapping, however, `TableSinkUtils` just extract the index.

Reply:
Just a minor suggestion.
I not mean share codes, just put all pojo things together is better to understand.